src/core/ctxt.ml

```diff
@@ -3,65 +3,57 @@
 open Extra
 open Terms
 
-(** Representation of a typing context, associating a type (or [Term.term]) to
-    free [Bindlib] variables. *)
-type t = (tvar * term) list
-
 (** [empty] is the empty context. *)
-let empty : t = []
-
-(** [add x a ctx] maps the variable [x] to the type [a] in [ctx]. *)
-let add : tvar -> term -> t -> t =
-  fun x a ctx -> (x,a)::ctx
-
-(** [unbind_ctxt ctx a b] returns the triple [(x,b,ctx')] such that [(x,b)]
-   is the unbinding of [b] and [ctx'] is the context [ctx] extended with
-   [(x,a)] if [x] occurs in [b]. *)
-let unbind ctx a b =
+let empty : ctxt = []
+
+(** [unbind ctx a def b] returns the triple [(x,b,ctx')] such that [(x,b)] is
+    the unbinding of [b] and [ctx'] is the context [ctx] extended with, if [x]
+    occurs in [b]
+    - the assumption that [x] is of type [a] if [def] is [None], and
+    - the definition of [x] in [def] of type [a] otherwise. *)
+let unbind : ctxt -> term -> term option -> tbinder -> tvar * term * ctxt =
+  fun ctx a def b ->
   let (x,b') = Bindlib.unbind b in
-  let ctx' = if Bindlib.binder_occur b then add x a ctx else ctx in
+  let ctx' = if Bindlib.binder_occur b then (x, a, def) :: ctx else ctx in
   (x,b',ctx')
```

Comment:
I know you tried to keep the changes minimal here, but could you please change the body of the function to the following:
```ocaml
let (x, t) = Bindlib.unbind b in
(x, t, if Bindlib.binder_occur b then (x, a, def) :: ctx else ctx)
```

Also, I don't like the comment very much. Please change it to:
```ocaml
(** [unbind ctx a def b] returns a triple [(x,t,new_ctx)] such that [(x,t)] is
    an unbinding of [b] (in the sense of [Bindlib.unbind]) and [new_ctx] is an
    extension of context [ctx] with the assumption that [x] has type [a] (only
    if [x] occurs in [t]). If [def] is of the form [Some(u)], the context also
    registers the term [u] as the definition of variable [x]. *)
```